src/nvim/api/vim.c

```diff
@@ -2389,3 +2389,153 @@ Dictionary nvim_complete_set(Integer index, Dict(complete_set) *opts, Arena *are
   }
   return rv;
 }
+
+enum { kStatusColumn = 0x1, kStatusLine = 0x2, kTabLine = 0x4, kWinBar = 0x8, };
+
+static void redraw_status(win_T *wp, buf_T *buf, int status, bool *flush)
+{
+  if (buf != NULL && wp->w_buffer != buf) {
+    return;
+  }
+
+  if (*wp->w_p_stc != NUL && (status & kStatusColumn)) {
+    wp->w_nrwidth_line_count = 0;
+    changed_window_setting(wp);
+  }
+  win_grid_alloc(wp);
+
+  // Just hid or showing winbar for the first time, or drawing the statuscolumn.
+  // Flush later because the window itself should be redrawn.
+  if (wp->w_lines_valid == 0) {
+    *flush = true;
+  }
+
+  if (*flush && (status & (kWinBar|kStatusLine))) {
+    // When flushing later, just mark for redraw.
+    wp->w_redr_status = true;
+  } else if (status & (kWinBar|kStatusLine)) {
+    // Not flushing later, only redraw the 'winbar' or 'statusline'.
+    win_check_ns_hl(wp);
+    if (status & kWinBar) {
+      win_redr_winbar(wp);
+    }
+    if (status & kStatusLine) {
+      win_redr_status(wp);
+    }
+    win_check_ns_hl(NULL);
+  }
+}
+
+/// Instruct Nvim to redraw various components.
+///
+/// @param opts  Optional parameters.
+///               - win: Target a specific |window-ID| for various components.
+///               - buf: Target a specific buffer number for various
+///                 components.
+///               - flush: Update the screen with pending updates, even if
+///                 Nvim is waiting for input and would normally not do so.
+///               - clear: Forcefully redraw `win`, all windows containing
+///                 `buf`, or all windows.
+///               - range: Buffer range to redraw, useful for decoration
+///                 providers. Tuple `[first, last]`:
+///                   - first: First line number of the range.
+///                   - last: Last line number of the range.
```

Comment:
There seems to be some line wrapping problems with docgen, plus there is no need to mention "line number of the range" twice
```suggestion
///                 providers. Tuple `[first, last]` are the first and last
///                 line numbers of the range, respectively.
```